ahocorasick.go

```diff
@@ -231,21 +231,44 @@ func (r Replacer) ReplaceAll(haystack string, replaceWith []string) string {
 
 type Finder interface {
 	FindAll(haystack string) []Match
+	FindN(haystack string, n int) []Match
 	PatternCount() int
 }
 
 // FindAll returns the matches found in the haystack
 func (ac AhoCorasick) FindAll(haystack string) []Match {
+	return ac.FindN(haystack, -1)
+}
+
+// FindN returns the matches found in the haystack.
+//
+// The count determines the number of matches to return:
+//   n > 0: at most n matches
+//   n == 0: the result is nil (zero matches)
+//   n < 0: all matches
+func (ac AhoCorasick) FindN(haystack string, n int) []Match {
+	if n == 0 {
+		return nil
+	}
+
 	iter := ac.Iter(haystack)
-	matches := make([]Match, 0)
+	var matches []Match
+	if n > 0 {
+		matches = make([]Match, 0, n)
+	}
 
+	i := 0
 	for {
 		next := iter.Next()
```

Comment:
It looks like using `ac.Iter` instead wouldn't be that much harder code but would be a bit more efficient for not `make([]Match)`. So I don't know if adding a `FindN` too is that helpful.

I'm not too sure what the intent of the `Finder` interface is, but should the iteration methods be added to it?

Reply:
Ah also noticed there is a `ReplaceAllFunc`, so if adding a method, following the same pattern `FindAllFunc` instead of `FindN` is probably more consistent in this code